### ISSUE TITLE
The roadmap has been sitting for a while.

### ISSUE THREAD

**Issue by bethelangela** [jzhang38/TinyLlama#155]
![image](https://github.com/jzhang38/TinyLlama/assets/159810751/0f0f5b9f-85c6-41d7-83ea-a5e6d78710c8)

Is there anything we can do with at least the first one?

I'd like to know how to train it on the New Testament as the case for pretraining in "other datasets". But I found the following. Maybe starting with [this KJV dataset](https://github.com/Jcharis/Bible-NLP-and-ML-using-Python/blob/master/kjv_cleandata.csv). So, I tried to play around it and the code is still running on colab. Hope it works.

I don't know if my code is right, but if it is, I'd like to see some implementation like this in this repo to update the roadmap.

Thank you.

```
import os
import pandas as pd
import json
import torch
import shutil
from os import system
from datetime import datetime
from datasets import load_dataset
from peft import LoraConfig
from peft.utils.other import prepare_model_for_kbit_training
from transformers import (
    AutoModelForCausalLM,
    AutoTokenizer,
    BitsAndBytesConfig,
    HfArgumentParser,
    AutoTokenizer,
    TrainingArguments
)

from trl import SFTTrainer

print("Loading the dataset...")
df = pd.read_csv("bible.csv")

# Select relevant columns and rename them
df = df[['id', 'chapter', 'verse', 'text']]
df.columns = ['label', 'chapter', 'verse', 'text']

# Concatenate 'label', 'chapter', and 'verse' into a single 'label' column
df['label'] = df['label'].astype(str) + ' ' + df['chapter'].astype(str) + ':' + df['verse'].astype(str)

# Select relevant columns
df = df[['label', 'text']]

# Convert DataFrame to JSON with the desired format
result = df.to_dict(orient='records')

#result = list(df.head(1000).to_json(orient="records"))
result = list( df.to_json( orient="records" ) )
result[0] = '{"json":['
result[-1] = ']'
result.append('}')
result = ''.join(result)
result = result.strip('"\'')
result = json.loads( result )

with open( 'data.json', 'w' ) as json_file:
    json.dump( result, json_file )

modelName = "TinyLlama/TinyLlama-1.1B-Chat-v1.0"
model = AutoModelForCausalLM.from_pretrained(
    modelName,
    device_map = 'auto',
    trust_remote_code = True,
    token = False,
)
model = prepare_model_for_kbit_training( model )

peft_config = LoraConfig(
    r = 32,
    lora_alpha = 16,
    bias = "none",
    lora_dropout = 0.05, # Conventional
    task_type = "CAUSAL_LM",
)
model.add_adapter( peft_config )
model.config.pretraining_tp = 1
tokenizer = AutoTokenizer.from_pretrained( modelName, trust_remote_code = True, use_fast=False )
tokenizer.pad_token = tokenizer.eos_token

if os.path.isdir( "./temp" ):
    shutil.rmtree( "./temp" )

model.config.use_cache = False

def formatting_func(example):
    text = f"<|system|></s>\n<|user|>Question: {example['label']}\n</s><|assistant|>Answer: {example['text']}</s>\n"
    return text

def generate_and_tokenize_prompt( prompt ):
    return tokenizer( formatting_func( prompt ), truncation = True, max_length = 2048 )

dataset = load_dataset("json", data_files="data.json", field='json', split="train")
dataset = dataset.map( generate_and_tokenize_prompt )

training_arguments = TrainingArguments(
        per_device_train_batch_size = 2,
        gradient_accumulation_steps = 4,
        warmup_ratio = 0.1,
        num_train_epochs = 1,
        learning_rate = 2e-5,
        fp16 = not torch.cuda.is_bf16_supported(),
        bf16 = torch.cuda.is_bf16_supported(),
        logging_steps = 1,
        optim = "adamw_8bit",
        weight_decay = 0.1,
        lr_scheduler_type = "linear",
        seed = 3407,
        output_dir = "outputs",
)

trainer = SFTTrainer(
    model=model,
    train_dataset=dataset,
    peft_config=peft_config,
    dataset_text_field="text",
    max_seq_length=2048,
    tokenizer=tokenizer,
    args=training_arguments,
    packing=False,
)

trainer.train()

if os.path.isdir( "./temp" ):
    shutil.rmtree( "./temp" )

directory = "trained"

if os.path.isdir( directory ):
    shutil.rmtree( directory )

model.save_pretrained( directory )
tokenizer.save_pretrained( directory )
print( f"Model saved '{directory}'." )
```

**Comment by oliverbob** [comment]
Tried your code on Colab. It took an hour to train.

TrainOutput(global_step=3888, training_loss=1.6469270712063637, metrics={'train_runtime': 1870.5777, 'train_samples_per_second': 16.627, 'train_steps_per_second': 2.079, 'total_flos': 8380872833605632.0, 'train_loss': 1.6469270712063637, 'epoch': 1.0})

I don't know if this is good result. 1.646927071206363 loss.
